New export API with dynamic shape specifications instead of constraints

avikchaudhuri opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #108448

Our experience using `constraints` / `dynamic_dim` with the existing export API has found it to be (subjectively) clunky and (objectively) verbose in common cases.

This PR implements a new design for the export API that replaces the use of `constraints` / `dynamic_dim` with a new way of specifying dynamic shapes, involving the following concepts: 
* a constructor `Dim` for first-class named dynamic dimensions with ranges (similar to `functorch.dim`, and analogous to internal symbolic sizes)
* a mechanism that uses the above in `export` calls to associate inputs to their dynamic shape specifications (`dynamic_shapes`)

Design doc: https://docs.google.com/presentation/d/168U7XK72C_WSsZpGESP6Cho9udh193fi0gfjxCNcJ4E/edit#slide=id.p (Meta-only). Note that we only implement Option 1 in that doc. An older version of this PR also implemented Option 3, which is an alternative way of specifying dynamic shapes using tensor type annotations on the exported callable; but we have moved that to future work for now.

See docs for these new features in `torch.export`. The existing `torch.export.export` is modified to use the new API, `torch._export.export__RC__`, whenever `constraints=None`. We have not deprecated the existing API yet, but will do in a follow-up.

Constraint violation errors arising through use of the new API will now contain suggested fixes using the new API. No longer do we need to report all specializations for static dimensions and suggest all constraints over dynamic dimensions to fix such errors. Instead, due to the redesign, the suggested fixes are much more concise, only involving modifying the definitions of relevant `Dim`s.

Differential Revision: [D48919204](https://our.internmc.facebook.com/intern/diff/D48919204/)

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

avikchaudhuri committed:
[RFC] alternative designs for export api



Differential Revision: [D48919204](https://our.internmc.facebook.com/intern/diff/D48919204/)

[ghstack-poisoned]

avikchaudhuri committed:
Update on "[RFC] alternative designs for export api"


Differential Revision: [D48919204](https://our.internmc.facebook.com/intern/diff/D48919204/)

[ghstack-poisoned]

avikchaudhuri committed:
Update on "[RFC] alternative designs for export api"


Differential Revision: [D48919204](https://our.internmc.facebook.com/intern/diff/D48919204/)

[ghstack-poisoned]

avikchaudhuri committed:
Update on "[RFC] alternative designs for export api"


Differential Revision: [D48919204](https://our.internmc.facebook.com/intern/diff/D48919204/)

[ghstack-poisoned]

avikchaudhuri committed:
Update on "[RFC] alternative designs for export api"


Differential Revision: [D48919204](https://our.internmc.facebook.com/intern/diff/D48919204/)

[ghstack-poisoned]

avikchaudhuri committed:
Update on "[RFC] alternative designs for export api"


Differential Revision: [D48919204](https://our.internmc.facebook.com/intern/diff/D48919204/)

[ghstack-poisoned]

avikchaudhuri committed:
Update on "[RFC] alternative designs for export api"


Differential Revision: [D48919204](https://our.internmc.facebook.com/intern/diff/D48919204/)

[ghstack-poisoned]

avikchaudhuri committed:
Update on "[RFC] alternative designs for export api"


Differential Revision: [D48919204](https://our.internmc.facebook.com/intern/diff/D48919204/)

[ghstack-poisoned]

avikchaudhuri committed:
Update on "[RFC] alternative designs for export api"


Differential Revision: [D48919204](https://our.internmc.facebook.com/intern/diff/D48919204/)

[ghstack-poisoned]

avikchaudhuri committed:
Update on "[RFC] alternative designs for export api"


Differential Revision: [D48919204](https://our.internmc.facebook.com/intern/diff/D48919204/)

[ghstack-poisoned]

avikchaudhuri committed:
Update on "[RFC] alternative designs for export api"


Differential Revision: [D48919204](https://our.internmc.facebook.com/intern/diff/D48919204/)

[ghstack-poisoned]

avikchaudhuri committed:
Update on "[RFC] alternative designs for export api"


Differential Revision: [D48919204](https://our.internmc.facebook.com/intern/diff/D48919204/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

avikchaudhuri committed:
Update on "[RFC] alternative designs for export api"


Differential Revision: [D48919204](https://our.internmc.facebook.com/intern/diff/D48919204/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

avikchaudhuri committed:
Update on "[RFC] alternative designs for export api"


Differential Revision: [D48919204](https://our.internmc.facebook.com/intern/diff/D48919204/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

avikchaudhuri committed:
Update on "[RFC] alternative designs for export api"


Differential Revision: [D48919204](https://our.internmc.facebook.com/intern/diff/D48919204/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

avikchaudhuri committed:
Update on "New export API with dynamic shape specifications instead of constraints"


Our experience using `constraints` / `dynamic_dim` with the existing export API has found it to be (subjectively) clunky and (objectively) verbose in common cases.

This PR implements a new design for the export API that replaces the use of `constraints` / `dynamic_dim` with a new way of specifying dynamic shapes, involving the following concepts: 
* a constructor `Dim` for first-class named dynamic dimensions with ranges (similar to `functorch.dim`, and analogous to internal symbolic sizes)
* a special value `_` denoting a static dimension
* specification mechanisms that use the above, for use either at an export call (`dynamic_shapes`) or on the export function itself (`TensorType`).

Design doc: https://docs.google.com/presentation/d/168U7XK72C_WSsZpGESP6Cho9udh193fi0gfjxCNcJ4E/edit#slide=id.p (Meta-only).

See docs for these new features in `torch._export`. The existing `torch.export.export` is modified to use the new API, `torch._export.export__RC__`, whenever `constraints=None`. We have not deprecated the existing API yet, but will do in a follow-up.

Constraint violation errors arising through use of the new API will now contain suggested fixes using the new API. No longer do we need to report all specializations for static dimensions and suggest all constraints over dynamic dimensions to fix such errors. Instead, due to the redesign, the suggested fixes are much more concise, only involving modifying the definitions of relevant `Dim`s.

Differential Revision: [D48919204](https://our.internmc.facebook.com/intern/diff/D48919204/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

avikchaudhuri committed:
Update on "New export API with dynamic shape specifications instead of constraints"


Our experience using `constraints` / `dynamic_dim` with the existing export API has found it to be (subjectively) clunky and (objectively) verbose in common cases.

This PR implements a new design for the export API that replaces the use of `constraints` / `dynamic_dim` with a new way of specifying dynamic shapes, involving the following concepts: 
* a constructor `Dim` for first-class named dynamic dimensions with ranges (similar to `functorch.dim`, and analogous to internal symbolic sizes)
* a special value `_` denoting a static dimension
* specification mechanisms that use the above, for use either at an export call (`dynamic_shapes`) or on the export function itself (`TensorType`).

Design doc: https://docs.google.com/presentation/d/168U7XK72C_WSsZpGESP6Cho9udh193fi0gfjxCNcJ4E/edit#slide=id.p (Meta-only).

See docs for these new features in `torch._export`. The existing `torch.export.export` is modified to use the new API, `torch._export.export__RC__`, whenever `constraints=None`. We have not deprecated the existing API yet, but will do in a follow-up.

Constraint violation errors arising through use of the new API will now contain suggested fixes using the new API. No longer do we need to report all specializations for static dimensions and suggest all constraints over dynamic dimensions to fix such errors. Instead, due to the redesign, the suggested fixes are much more concise, only involving modifying the definitions of relevant `Dim`s.

Differential Revision: [D48919204](https://our.internmc.facebook.com/intern/diff/D48919204/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

gmagogsfm commented on torch/_export/__init__.py:
```diff
@@ -62,7 +65,157 @@
+    Note that different dimensions of the same tensor, or of different tensors,
+    can be described by the same type.
+    """
+    _min = 2 if min is None else min
+    _max = sys.maxsize if max is None else max
```
```suggestion
    _min = 2 if min is None else builtins.max(min, 2)
    _max = sys.maxsize if max is None else builtins.max(max, 2)
    assert _max > _min
```

gmagogsfm commented on torch/_export/__init__.py:
```diff
@@ -62,7 +65,157 @@
+    return _Dim(name, (int,), {"min": _min, "max": _max})
+
+
+def dims(*names: str):
```
Just making sure I understand this correctly: users are not supposed to modify min/max of created Dim objects, and dims doesn't take max/min arguments, so `dims` can only be used to attach names to dims?

avikchaudhuri replied:
Sorry, this is an oversight. I should add min/max args here.

avikchaudhuri commented on torch/export/__init__.py:
```diff
@@ -693,12 +705,18 @@ def __eq__(self, other):
+        if self.debug_name is None:
+            debug_name = other.debug_name
+        else:
+            assert other.debug_name is None or self.debug_name == other.debug_name
```
This only guards against inconsistent debug names. With the old API debug names will always be `None`.

avikchaudhuri commented on torch/_export/__init__.py:
```diff
@@ -62,7 +65,157 @@
+    signature = inspect.signature(f.forward) if isinstance(f, torch.nn.Module) else inspect.signature(f)
+    combined_args = signature.bind(*args, **kwargs).arguments
+
+    if dynamic_shapes is None:
```
Yes.

gmagogsfm replied:
Should this part of logic be removed now that we don't have TensorType?

avikchaudhuri replied:
Oh yes.

avikchaudhuri commented on torch/export/__init__.py:
```diff
@@ -1020,9 +1042,12 @@ def specify_constraints(x):
+    from torch._export import export, export__RC__
 
-    return export(f, args, kwargs, constraints)
+    if constraints is not None:
+        return export(f, args, kwargs, constraints)
+    else:
+        return export__RC__(f, args, kwargs, dynamic_shapes=dynamic_shapes)
```
I'll punt on it. Otherwise I'll have to check annotations right here. Given we will deprecate or even remove `constraints` soon, I doubt this will come up in practice.

avikchaudhuri commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1960,7 +1966,103 @@ def remove_redundant_dynamic_results(self):
+                left, op, right = t.split(" ")
+                if op == "==" and left == right:
+                    continue
+                if left.isdigit():
```
`op` needs to be flipped too

gmagogsfm replied:
Yeah, replacing all "<" with ">" and vice versa should be enough.

avikchaudhuri committed:
Update on "New export API with dynamic shape specifications instead of constraints"


Our experience using `constraints` / `dynamic_dim` with the existing export API has found it to be (subjectively) clunky and (objectively) verbose in common cases.

This PR implements a new design for the export API that replaces the use of `constraints` / `dynamic_dim` with a new way of specifying dynamic shapes, involving the following concepts: 
* a constructor `Dim` for first-class named dynamic dimensions with ranges (similar to `functorch.dim`, and analogous to internal symbolic sizes)
* a special value `_` denoting a static dimension
* specification mechanisms that use the above, for use either at an export call (`dynamic_shapes`) or on the export function itself (`TensorType`).

Design doc: https://docs.google.com/presentation/d/168U7XK72C_WSsZpGESP6Cho9udh193fi0gfjxCNcJ4E/edit#slide=id.p (Meta-only).

See docs for these new features in `torch._export`. The existing `torch.export.export` is modified to use the new API, `torch._export.export__RC__`, whenever `constraints=None`. We have not deprecated the existing API yet, but will do in a follow-up.

Constraint violation errors arising through use of the new API will now contain suggested fixes using the new API. No longer do we need to report all specializations for static dimensions and suggest all constraints over dynamic dimensions to fix such errors. Instead, due to the redesign, the suggested fixes are much more concise, only involving modifying the definitions of relevant `Dim`s.

Differential Revision: [D48919204](https://our.internmc.facebook.com/intern/diff/D48919204/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

avikchaudhuri committed:
Update on "New export API with dynamic shape specifications instead of constraints"


Our experience using `constraints` / `dynamic_dim` with the existing export API has found it to be (subjectively) clunky and (objectively) verbose in common cases.

This PR implements a new design for the export API that replaces the use of `constraints` / `dynamic_dim` with a new way of specifying dynamic shapes, involving the following concepts: 
* a constructor `Dim` for first-class named dynamic dimensions with ranges (similar to `functorch.dim`, and analogous to internal symbolic sizes)
* a special value `_` denoting a static dimension
* specification mechanisms that use the above, for use either at an export call (`dynamic_shapes`) or on the export function itself (`TensorType`).

Design doc: https://docs.google.com/presentation/d/168U7XK72C_WSsZpGESP6Cho9udh193fi0gfjxCNcJ4E/edit#slide=id.p (Meta-only).

See docs for these new features in `torch._export`. The existing `torch.export.export` is modified to use the new API, `torch._export.export__RC__`, whenever `constraints=None`. We have not deprecated the existing API yet, but will do in a follow-up.

Constraint violation errors arising through use of the new API will now contain suggested fixes using the new API. No longer do we need to report all specializations for static dimensions and suggest all constraints over dynamic dimensions to fix such errors. Instead, due to the redesign, the suggested fixes are much more concise, only involving modifying the definitions of relevant `Dim`s.

Differential Revision: [D48919204](https://our.internmc.facebook.com/intern/diff/D48919204/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

avikchaudhuri commented on torch/_export/__init__.py:
```diff
@@ -62,7 +66,204 @@
+    - Arguments that are dicts or tuples of tensors are recursively specified
+      by using mappings or sequences of contained specifications.
+
+    Mechanism #2: types
```
> First, there should ideally be one way to do things. The type annotation mechanism seems strictly less powerful (e.g. it seems impossible to express something like a dict where specific keys have different constraints). I think if we want two mechanisms, there needs to be a strong reason.

I agree that one way is better than two, but I think the call-site and the function-site use cases cater to different sets of users.

Actually there is enough power in the type annotation mechanism right now to match whatever you can pass to `dynamic_shapes`. In particular I've included a `dynamic_shapes_type` function to map one to the other. For the case you mention, you'd use a `TypedDict`.

That said I'm somewhat frustrated that a lot of good things in `typing` happened only in 3.9+ and so I've to resort to less-than-ideal implementations to remain compatible with 3.8.

> Second, static type annotations come with the user expectation that the annotation is universally true, e.g. if I type a function def foo(i: int), I come in with the expectation that i will always be an int, regardless of how foo is called. This is not true of these type annotations—they are only respected if the function happens to be the top-level exported function. I can see this being potentially quite confusing.

I agree. Coupled with the fact that I'm not quite happy with the way the type annotations are encoded right now, this feels like a good reason to take it out.

I do absolutely want to bring it back at some point. So I might have to leave in some tests to ensure that in principle the two mechanisms are equally expressive.

avikchaudhuri committed:
Update on "New export API with dynamic shape specifications instead of constraints"


Our experience using `constraints` / `dynamic_dim` with the existing export API has found it to be (subjectively) clunky and (objectively) verbose in common cases.

This PR implements a new design for the export API that replaces the use of `constraints` / `dynamic_dim` with a new way of specifying dynamic shapes, involving the following concepts: 
* a constructor `Dim` for first-class named dynamic dimensions with ranges (similar to `functorch.dim`, and analogous to internal symbolic sizes)
* a special value `_` denoting a static dimension
* specification mechanisms that use the above, for use either at an export call (`dynamic_shapes`) or on the export function itself (`TensorType`).

Design doc: https://docs.google.com/presentation/d/168U7XK72C_WSsZpGESP6Cho9udh193fi0gfjxCNcJ4E/edit#slide=id.p (Meta-only).

See docs for these new features in `torch._export`. The existing `torch.export.export` is modified to use the new API, `torch._export.export__RC__`, whenever `constraints=None`. We have not deprecated the existing API yet, but will do in a follow-up.

Constraint violation errors arising through use of the new API will now contain suggested fixes using the new API. No longer do we need to report all specializations for static dimensions and suggest all constraints over dynamic dimensions to fix such errors. Instead, due to the redesign, the suggested fixes are much more concise, only involving modifying the definitions of relevant `Dim`s.

Differential Revision: [D48919204](https://our.internmc.facebook.com/intern/diff/D48919204/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

avikchaudhuri committed:
Update on "New export API with dynamic shape specifications instead of constraints"


Our experience using `constraints` / `dynamic_dim` with the existing export API has found it to be (subjectively) clunky and (objectively) verbose in common cases.

This PR implements a new design for the export API that replaces the use of `constraints` / `dynamic_dim` with a new way of specifying dynamic shapes, involving the following concepts: 
* a constructor `Dim` for first-class named dynamic dimensions with ranges (similar to `functorch.dim`, and analogous to internal symbolic sizes)
* a special value `_` denoting a static dimension
* specification mechanisms that use the above, for use either at an export call (`dynamic_shapes`) or on the export function itself (`TensorType`).

Design doc: https://docs.google.com/presentation/d/168U7XK72C_WSsZpGESP6Cho9udh193fi0gfjxCNcJ4E/edit#slide=id.p (Meta-only).

See docs for these new features in `torch._export`. The existing `torch.export.export` is modified to use the new API, `torch._export.export__RC__`, whenever `constraints=None`. We have not deprecated the existing API yet, but will do in a follow-up.

Constraint violation errors arising through use of the new API will now contain suggested fixes using the new API. No longer do we need to report all specializations for static dimensions and suggest all constraints over dynamic dimensions to fix such errors. Instead, due to the redesign, the suggested fixes are much more concise, only involving modifying the definitions of relevant `Dim`s.

Differential Revision: [D48919204](https://our.internmc.facebook.com/intern/diff/D48919204/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

avikchaudhuri commented on torch/_export/__init__.py:
```diff
@@ -62,7 +66,204 @@
+    from collections.abc import Mapping, Sequence
+    from typing import get_origin, get_args, _TypedDictMeta  # type: ignore[attr-defined]
+
+    def assoc_zip(combined_args, dynamic_shapes):
```
Handling dataclasses in the most recent version. 

I would have used `pytree` if there was something like `tree_zip` (something that descends into the same structure of multiple things at once). I think having this custom descent is fine for now.

Key paths: yeah, maybe good for partial specs later. For now it is convenient enough to `pytree.tree_map` over inputs to derive specs.

suo replied:
yeah agree, it's just an implementation detail. We can improve in the future

gmagogsfm reviewed (approved):
Looks good, just some minor nits

avikchaudhuri committed:
Update on "New export API with dynamic shape specifications instead of constraints"


Our experience using `constraints` / `dynamic_dim` with the existing export API has found it to be (subjectively) clunky and (objectively) verbose in common cases.

This PR implements a new design for the export API that replaces the use of `constraints` / `dynamic_dim` with a new way of specifying dynamic shapes, involving the following concepts: 
* a constructor `Dim` for first-class named dynamic dimensions with ranges (similar to `functorch.dim`, and analogous to internal symbolic sizes)
* a special value `_` denoting a static dimension
* specification mechanisms that use the above, for use either at an export call (`dynamic_shapes`) or on the export function itself (`TensorType`).

Design doc: https://docs.google.com/presentation/d/168U7XK72C_WSsZpGESP6Cho9udh193fi0gfjxCNcJ4E/edit#slide=id.p (Meta-only).

See docs for these new features in `torch._export`. The existing `torch.export.export` is modified to use the new API, `torch._export.export__RC__`, whenever `constraints=None`. We have not deprecated the existing API yet, but will do in a follow-up.

Constraint violation errors arising through use of the new API will now contain suggested fixes using the new API. No longer do we need to report all specializations for static dimensions and suggest all constraints over dynamic dimensions to fix such errors. Instead, due to the redesign, the suggested fixes are much more concise, only involving modifying the definitions of relevant `Dim`s.

Differential Revision: [D48919204](https://our.internmc.facebook.com/intern/diff/D48919204/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

suo reviewed (approved):
lgtm!

avikchaudhuri committed:
Update on "New export API with dynamic shape specifications instead of constraints"


Our experience using `constraints` / `dynamic_dim` with the existing export API has found it to be (subjectively) clunky and (objectively) verbose in common cases.

This PR implements a new design for the export API that replaces the use of `constraints` / `dynamic_dim` with a new way of specifying dynamic shapes, involving the following concepts: 
* a constructor `Dim` for first-class named dynamic dimensions with ranges (similar to `functorch.dim`, and analogous to internal symbolic sizes)
* a special value `_` denoting a static dimension
* specification mechanisms that use the above, for use either at an export call (`dynamic_shapes`) or on the export function itself (`TensorType`).

Design doc: https://docs.google.com/presentation/d/168U7XK72C_WSsZpGESP6Cho9udh193fi0gfjxCNcJ4E/edit#slide=id.p (Meta-only).

See docs for these new features in `torch._export`. The existing `torch.export.export` is modified to use the new API, `torch._export.export__RC__`, whenever `constraints=None`. We have not deprecated the existing API yet, but will do in a follow-up.

Constraint violation errors arising through use of the new API will now contain suggested fixes using the new API. No longer do we need to report all specializations for static dimensions and suggest all constraints over dynamic dimensions to fix such errors. Instead, due to the redesign, the suggested fixes are much more concise, only involving modifying the definitions of relevant `Dim`s.

Differential Revision: [D48919204](https://our.internmc.facebook.com/intern/diff/D48919204/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

avikchaudhuri committed:
Update on "New export API with dynamic shape specifications instead of constraints"


Our experience using `constraints` / `dynamic_dim` with the existing export API has found it to be (subjectively) clunky and (objectively) verbose in common cases.

This PR implements a new design for the export API that replaces the use of `constraints` / `dynamic_dim` with a new way of specifying dynamic shapes, involving the following concepts: 
* a constructor `Dim` for first-class named dynamic dimensions with ranges (similar to `functorch.dim`, and analogous to internal symbolic sizes)
* a mechanism that uses the above in `export` calls to associate inputs to their dynamic shape specifications (`dynamic_shapes`)

Design doc: https://docs.google.com/presentation/d/168U7XK72C_WSsZpGESP6Cho9udh193fi0gfjxCNcJ4E/edit#slide=id.p (Meta-only). Note that we only implement Option 1 in that doc. An older version of this PR also implemented Option 3, which is an alternative way of specifying dynamic shapes using tensor type annotations on the exported callable; but we have moved that to future work for now.

See docs for these new features in `torch.export`. The existing `torch.export.export` is modified to use the new API, `torch._export.export__RC__`, whenever `constraints=None`. We have not deprecated the existing API yet, but will do in a follow-up.

Constraint violation errors arising through use of the new API will now contain suggested fixes using the new API. No longer do we need to report all specializations for static dimensions and suggest all constraints over dynamic dimensions to fix such errors. Instead, due to the redesign, the suggested fixes are much more concise, only involving modifying the definitions of relevant `Dim`s.

Differential Revision: [D48919204](https://our.internmc.facebook.com/intern/diff/D48919204/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

avikchaudhuri committed:
Update on "New export API with dynamic shape specifications instead of constraints"


Our experience using `constraints` / `dynamic_dim` with the existing export API has found it to be (subjectively) clunky and (objectively) verbose in common cases.

This PR implements a new design for the export API that replaces the use of `constraints` / `dynamic_dim` with a new way of specifying dynamic shapes, involving the following concepts: 
* a constructor `Dim` for first-class named dynamic dimensions with ranges (similar to `functorch.dim`, and analogous to internal symbolic sizes)
* a mechanism that uses the above in `export` calls to associate inputs to their dynamic shape specifications (`dynamic_shapes`)

Design doc: https://docs.google.com/presentation/d/168U7XK72C_WSsZpGESP6Cho9udh193fi0gfjxCNcJ4E/edit#slide=id.p (Meta-only). Note that we only implement Option 1 in that doc. An older version of this PR also implemented Option 3, which is an alternative way of specifying dynamic shapes using tensor type annotations on the exported callable; but we have moved that to future work for now.

See docs for these new features in `torch.export`. The existing `torch.export.export` is modified to use the new API, `torch._export.export__RC__`, whenever `constraints=None`. We have not deprecated the existing API yet, but will do in a follow-up.

Constraint violation errors arising through use of the new API will now contain suggested fixes using the new API. No longer do we need to report all specializations for static dimensions and suggest all constraints over dynamic dimensions to fix such errors. Instead, due to the redesign, the suggested fixes are much more concise, only involving modifying the definitions of relevant `Dim`s.

Differential Revision: [D48919204](https://our.internmc.facebook.com/intern/diff/D48919204/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

avikchaudhuri committed:
Update on "New export API with dynamic shape specifications instead of constraints"


Our experience using `constraints` / `dynamic_dim` with the existing export API has found it to be (subjectively) clunky and (objectively) verbose in common cases.

This PR implements a new design for the export API that replaces the use of `constraints` / `dynamic_dim` with a new way of specifying dynamic shapes, involving the following concepts: 
* a constructor `Dim` for first-class named dynamic dimensions with ranges (similar to `functorch.dim`, and analogous to internal symbolic sizes)
* a mechanism that uses the above in `export` calls to associate inputs to their dynamic shape specifications (`dynamic_shapes`)

Design doc: https://docs.google.com/presentation/d/168U7XK72C_WSsZpGESP6Cho9udh193fi0gfjxCNcJ4E/edit#slide=id.p (Meta-only). Note that we only implement Option 1 in that doc. An older version of this PR also implemented Option 3, which is an alternative way of specifying dynamic shapes using tensor type annotations on the exported callable; but we have moved that to future work for now.

See docs for these new features in `torch.export`. The existing `torch.export.export` is modified to use the new API, `torch._export.export__RC__`, whenever `constraints=None`. We have not deprecated the existing API yet, but will do in a follow-up.

Constraint violation errors arising through use of the new API will now contain suggested fixes using the new API. No longer do we need to report all specializations for static dimensions and suggest all constraints over dynamic dimensions to fix such errors. Instead, due to the redesign, the suggested fixes are much more concise, only involving modifying the definitions of relevant `Dim`s.

Differential Revision: [D48919204](https://our.internmc.facebook.com/intern/diff/D48919204/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

avikchaudhuri commented:
@pytorchbot merge

avikchaudhuri commented:
@pytorchbot merge -f "unrelated CI failure"
